gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/CompareType.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.process.traversal;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * {@code CompareType} is a {@code BiPredicate} that determines whether the first argument is a type of the second argument.
+ *
+ */
+public enum CompareType implements PBiPredicate<Object, Class<?>> {
+
+    /**
+     * Evaluates if the first object is an instance of the second class.
+     *
+     * @since 3.8.0
+     */
+    typeOf {
+        @Override
+        public boolean test(final Object first, final Class<?> second) {
+            if (first == null) {
+                return second == null;
+            }
+            return second != null && second.isAssignableFrom(first.getClass());
+        }
+
+    };
+
+    public static final class GlobalTypeCache {
+
+        private GlobalTypeCache() {
+            throw new IllegalStateException("Utility class");
+        }
+
+        /**
+         * A register of enum names for all types.
+         */
+        private static final Map<String, Class<?>> GLOBAL_TYPE_REGISTRY = new ConcurrentHashMap<>();
+
+        // register the GType classes for convinience use of simple class names in grammar
+        static {
+            for (GType value : GType.values()) {
+                if (value == GType.NULL) {
+                    // skip the null type
+                    continue;
+                }
+                registerDataType(value.getType());
+            }
+        }
+
+        /**
+         * Register a type with a custom name.
```

Comment:
While I don't think it's our place to enforce any rules on name registration, would we want to define a convention on capitalization which we recommend providers follow? The preloaded types from GType appear to be taking the `PascalCase` names from their Java classes.

Reply:
let's be sure that these conventions make it to the provider documentation.